src/main/java/uk/gov/hmcts/reform/hmc/service/AccessControlServiceImpl.java

```diff
@@ -0,0 +1,169 @@
+import static uk.gov.hmcts.reform.hmc.repository.DefaultRoleAssignmentRepository.ROLE_ASSIGNMENT_INVALID_STATUS;
+import static uk.gov.hmcts.reform.hmc.repository.DefaultRoleAssignmentRepository.ROLE_ASSIGNMENT_MISSING_REQUIRED;
+
+@Service
+@Slf4j
+public class AccessControlServiceImpl implements AccessControlService {
+
+    private final RoleAssignmentService roleAssignmentService;
+    private final SecurityUtils securityUtils;
+    private final DataStoreRepository dataStoreRepository;
+    private final CaseHearingRequestRepository caseHearingRequestRepository;
+    private final HearingRepository hearingRepository;
+
+    public static final String HEARING_MANAGER = "hearing-manager";
+    public static final String HEARING_VIEWER = "hearing-viewer";
+    public static final String LISTED_HEARING_VIEWER = "listed-hearing-viewer";
+
+    private static final List<String> HMC_ROLE_NAMES = Lists.newArrayList(
+        HEARING_MANAGER,
+        HEARING_VIEWER,
+        LISTED_HEARING_VIEWER);
+
+    public AccessControlServiceImpl(RoleAssignmentService roleAssignmentService,
+                                    SecurityUtils securityUtils,
+                                    @Qualifier("defaultDataStoreRepository")
+                                        DataStoreRepository dataStoreRepository,
+                                    CaseHearingRequestRepository caseHearingRequestRepository,
+                                    HearingRepository hearingRepository) {
+        this.roleAssignmentService = roleAssignmentService;
+        this.securityUtils = securityUtils;
+        this.dataStoreRepository = dataStoreRepository;
+        this.caseHearingRequestRepository = caseHearingRequestRepository;
+        this.hearingRepository = hearingRepository;
+    }
+
+    @Override
+    public void verifyAccess(Long hearingId, List<String> requiredRoles) {
+        List<RoleAssignment> filteredRoleAssignments = verifyRoleAccess(requiredRoles, hearingId);
+        verifyHearingStatus(filteredRoleAssignments, null, hearingId);
+    }
+
+    @Override
+    public void verifyCaseAccess(String caseReference, List<String> requiredRoles) {
+        verifyCaseAccess(caseReference, requiredRoles, null);
+    }
+
+    public void verifyCaseAccess(String caseReference, List<String> requiredRoles, Long hearingId) {
+        List<RoleAssignment> filteredRoleAssignments = verifyRoleAccess(requiredRoles, hearingId);
+
+        DataStoreCaseDetails caseDetails = dataStoreRepository.findCaseByCaseIdUsingExternalApi(caseReference);
+        if (!checkRoleAssignmentMatchesCaseDetails(caseDetails, filteredRoleAssignments)) {
+            throw new InvalidRoleAssignmentException(ROLE_ASSIGNMENT_INVALID_ATTRIBUTES);
+        }
+        verifyHearingStatus(filteredRoleAssignments, caseDetails.getCaseTypeId(), hearingId);
+    }
+
+    private List<RoleAssignment> verifyRoleAccess(List<String> requiredRoles, Long hearingId) {
```

Comment:
```suggestion
    private List<RoleAssignment> verifyRoleAccess(List<String> requiredRoles) {
```
`hearingId` no longer used in this method